### PR TITLE
Remove dist object

### PR DESCRIPTION
**Pull request by kalefranz**

### CODE REVIEW

File: conda/cli/install.py [hunk 1]
```diff
@@ -145,8 +145,8 @@ def install(args, parser, command='install'):
 # $ conda update --prefix %s anaconda
 """ % prefix)
 
-    linked_dists = install_linked(prefix)
-    linked_names = tuple(ld.quad[0] for ld in linked_dists)
```
Comment (mcg1969) [inline conda/cli/install.py]: OK, this concerns me. We're doing all this work to get rid of Dist objects, but it's not addressing the core issue, in my view: that we're depending on the _key_ to extract things like name, version, build.

File: conda/cli/main_list.py [hunk 2]
```diff
@@ -113,7 +114,7 @@ def print_export_header():
 
 def get_packages(installed, regex):
     pat = re.compile(regex, re.I) if regex else None
-    for dist in sorted(installed, key=lambda x: x.quad[0].lower()):
+    for dist in sorted(installed, key=lambda x: Dist(x).quad[0].lower()):
```
Comment (mcg1969) [inline conda/cli/main_list.py]: This concerns me: we're jumping through all these hoops to get rid of Dist, but this `quad` monstrosity (of my creation) is still here.

File: conda/models/dist.py [hunk 3]
```diff
@@ -36,41 +41,61 @@ def __call__(cls, *args, **kwargs):
...
-                                   name=name,
-                                   version=version,
-                                   build_string=build_string,
-                                   build_number=build_number,
-                                   with_features_depends=with_features_depends,
-                                   base_url=base_url,
-                                   platform=platform)
+class Dist(object):
+
+    def __init__(self, channel, name=None, version=None, build_string=None, build_number=None,
```
Comment (mcg1969) [inline conda/models/dist.py]: This is going in the _opposite_ direction from the way we should be going. This is _increasing_ the dependence on the Dist object to determine name/version/build. We should be moving _away_ from that. Dist should _not_ have `name`, `version`, and `build_string` in it at all, and we _certainly_ should not be constructing the filename from it. The filename is what it is, and it not ought be required to follow this convention.

**Comment by kalefranz** [comment]
This is y ready to dig in yet!!!! Still WIP :)

I'm in the middle of things. Needed to go get food though, so figured I'd see what full CI did against it.

**Comment by mcg1969** [comment]
Well, I'd like to save you some trouble, because you're striking off in what I consider to be the wrong direction. Removing `dist_name` in favor of `name`/`version`/`build` is not the way to go. And as you know I've tried to do similar PRs for awhile now.

**Comment by kalefranz** [comment]
Oh I see what you mean. And I see your point now. I wasn't too concerned about it because I was really aiming toward getting rid of it altogether. But yeah I see what youresaying.

**Comment by mcg1969** [comment]
Yeah, I do think it will be harder to work this way. In fact, forcing yourself to _avoid_ using Dist objects for name, version, build determination will go a long way to reduce its importance.

**Comment by mcg1969** [comment]
Look, I really don't think this should be merged into 4.3.x. Changes to 4.3.x should be about specific bugs and targeted performance fixes, not something this comprehensive. You seem be going back and forth between wanting to merge this and worrying about breaking things---well, trust the worry.

#4476 is lower in scope, but I would say that unless we see a significant performance gain from either of these---and with the pickled PR, I doubt we will---then I think we should hold off.

**Comment by goanpeca** [comment]
@kalefranz lets create an issue associated to the work in this PR, so we start organizing things better :-)

**Comment by kalefranz** [comment]
Closing, as conda API work will probably cover this.

**Pull request closed without merging**